File: HTML-JS/sudokuSolver.js
/*** Program begins here when 'solve' button is pressed ***/
function readInput(inputForm){
	// read input fields to 2 9x9 matrices
	var inputMatrix = new Array(9);
	var solutionMatrix = new Array(9);
	for(var i = 0; i < inputMatrix.length; i++){
		inputMatrix[i] = new Array(9);
		solutionMatrix[i] = new Array(9);
		for(var j = 0; j < 9; j++){
			// convert from 1d array to 2d matrix coordinates
			var k = i*9 + j;
			// extract input field value
			var x = inputForm[k].value;
			// convert blank fields into 0's
			if(x == ""){
				x = 0;
			}
			// stop and alert if input is invalid
			if(isNaN(x) || x < 0 || x > 9){
				alert("Enter only numbers between 1 and 9\nCell("+(i+1)+","+(j+1)+"): "+x);
				return;
			}
			inputMatrix[i][j] = x;
			solutionMatrix[i][j] = x;
		}
	}
	console.log("Input matrix creation successful");

	// check if puzzle is solvable
	var valid = validateMatrix(inputMatrix);
	if(!valid){
		alert("Input isn't a valid puzzle");
		return;
	}
	console.log("Validation Passed");
	
	// read radio buttons and slider
	var realtime;
	var delay;
	if(inputForm.radio.value == "radio-realtime"){
		realtime = true;
		delay = inputForm.slider.value;
	}
	else{
		realtime = false;
		delay = 0;
	}
	
	// hide instructions button
	document.getElementById("inst-button").innerHTML = "";
	
	// solve the puzzle
	outputToTable(solutionMatrix,inputMatrix);
	solvePuzzle(inputMatrix,solutionMatrix,realtime,delay);
	// 'instant' mode needs to manually output result
	if(!realtime){
		outputToTable(solutionMatrix,inputMatrix);
	}
}
/*** Program terminates here ***/

// output matrix to html table
function outputToTable(solutionMatrix,inputMatrix){
	var output = "";
	output += "<table id=\"outputTable\">";
	for(var i = 0; i < 9; i++){
		output += "<tr>";
		for(var j = 0; j < 9; j++){
			output += "<td>"
			// identify numbers from original input
			if(inputMatrix[i][j] != 0){
				output += "<span id=\"original\">"+solutionMatrix[i][j]+"</span>";
			}
			// don't display 0's
			else if(solutionMatrix[i][j] == 0){
				output += "";
			}
			else{
				output += solutionMatrix[i][j];
			}
			output += "</td>";
		}
		output += "</tr>";
	}
	output += "</table>";
	document.getElementById("startForm").innerHTML = output;
}

// check if user entered a valid puzzle
function validateMatrix(matrix){
	for(var i = 0; i < matrix.length; i++) {
		for(var j = 0; j < matrix[0].length; j++) {
			if(matrix[i][j] < 0 || matrix[i][j] > 9) {
				return false;
			}
			if(matrix[i][j] != 0) {
				if(!testCell(matrix, i, j, matrix[i][j], false)) {
					return false;
				}
			}
		}
	}
	return true;
}

// test if given value can be added into given cell
function testCell(matrix,row,column,value,solvingCheck){
	// boolean 'solvingCheck' is true when solving and false when validating
	// it is used to skip the current cell during validation
	// test row and column
	for(var i = 0; i < 9; i++){
		if(i == column && !solvingCheck){
			continue;
		}
		else if(i == row && !solvingCheck) {
			continue;
		}
		else if(matrix[row][i] == value){
			return false;
		}
		else if(matrix[i][column] == value){
			return false;
		}
	}
	// test neighborhood 
	var xPos = Math.floor(row / 3);
	var yPos = Math.floor(column / 3);
	for(var i = xPos*3; i < (xPos*3)+3; i++){
		for(var j = yPos*3; j < (yPos*3)+3; j++){
			if(i == row && j == column){
				continue;
			}
			else if(matrix[i][j] == value){
				return false
			}
		}
	}
	// if row, column, and neighborhood passed
	return true;
}

// solve the puzzle
async function solvePuzzle(inputMatrix,solutionMatrix,realtime,delay){
	var i = 0; //row
	var j = 0; //column
	var moves = 0; //tracks number of attempted moves
	var forwards = true; //used to determine if program is moving to next cell or backtracking
	while(i < 9){
		// Original cell is not blank
		if(inputMatrix[i][j] != 0){
			// if moving forward, go to next cell
			if(forwards){
				if(j == 8){
					i++;
					j = 0;
				}else{
					j++;
				}
			}
			// if moving backward, go to previous cell
			else{
				if(j == 0){
					i--;
					j = 8;
				}else{
					j--;
				}
			}
		}
		// Original cell is blank
		else{
			var k = 1;
			for(k = solutionMatrix[i][j]; k <=9; k++){
				if(k == 0){
					//current value is 0, skip to 1 and start checking
					continue;
				}
				// Is k a valid entry
				if(testCell(solutionMatrix,i,j,k,true)){
					// set new entry value
					solutionMatrix[i][j] = k;
					forwards = true;
					moves++;
					// if realtime is true, the table will update each new entry and pause 'delay'ms
					if(realtime){
						var temp = document.getElementById("outputTable");
						temp.rows[i].cells[j].innerHTML = k;
						await sleep(delay);
					}
					break;
				}
			}
			// if loop stopped before reaching end, entry is good, go to next cell
			if(k < 10){
					if(j == 8){
						i++;
						j = 0;
					}else{
						j++;
					}
					continue;
			}
			// if loop ran all the way unsuccessfully
			else{
				// start working backwards
				forwards = false;
				// clear current cell value
				solutionMatrix[i][j] = 0;
				// go to previous cell
				if(j == 0){
					i--;
					j = 8;
				}else{
					j--;
				}
			}
		}
	}
	// insert commas into 'moves'
	var movesFormatted = new Intl.NumberFormat().format(moves);
	console.log("Completed in "+movesFormatted+" moves");
	document.getElementById("heading").innerHTML = "<h2>Solution</h2><p>Solved in "+movesFormatted+" moves</p>";
}

// used to add a delay when updating the table in realtime
function sleep(milliseconds){
	return new Promise(resolve => setTimeout(resolve, milliseconds));
}

File: Python/SudokuSolver.py
def main():
    #create input puzzle
    puzzle_matrix = [
        [5,3,0,0,7,0,0,0,0],
        [6,0,0,1,9,5,0,0,0],
        [0,9,8,0,0,0,0,6,0],
        [8,0,0,0,6,0,0,0,3],
        [4,0,0,8,0,3,0,0,1],
        [7,0,0,0,2,0,0,0,6],
        [0,6,0,0,0,0,2,8,0],
        [0,0,0,4,1,9,0,0,5],
        [0,0,0,0,8,0,0,7,9]]
    #copy input puzzle to working puzzle
    working_matrix = [[0 for x in range(9)] for y in range(9)]
    for i in range(len(puzzle_matrix)):
        for j in range(len(puzzle_matrix[0])):
            working_matrix[i][j] = puzzle_matrix[i][j]
    i = 0
    j = 0
    moves = 0
    forwards = True #used to determine if program is moving to next cell or backtracking
    while i < 9:
        #puzzle cell is not blank
        if puzzle_matrix[i][j] != 0:
            #'forward' determines if moving forwards or backwards
            i,j = move_cell(i, j, forwards)
        #puzzle cell is blank
        else:
            valid = False   #used to determine if entry was set
            k = working_matrix[i][j]
            for k in range(k,10):
                if k == 0:
                    #current value is 0, skip to 1 and start checking
                    continue
                #is k a valid entry?
                test = test_cell(working_matrix, i, j, k, True)
                if test == True:
                    #set new entry value
                    working_matrix[i][j] = k
                    forwards = True
                    moves = moves + 1
                    valid = True
                    break
            #if entry is valid - go to next cell
            #if entry wasn't valid - start working backwards, clear current cell, go to previous cell
            if valid == False:
                forwards = False
                working_matrix[i][j] = 0
            i,j = move_cell(i, j, valid)
    print_matrix(working_matrix)
    print("\nSolved in "+str(moves)+" moves")
    return

def move_cell(i, j, test):
    if test == True:
        if j == 8:
            i = i + 1
            j = 0
        else:
            j = j + 1
    else:
        if j == 0:
            i = i - 1
            j = 8
        else:
            j = j - 1
    return i,j

def validate_matrix(matrix):
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            if matrix[i][j] < 0 or matrix[i][j] > 9:
                return False
            if matrix[i][j] != 0:
                test = test_cell(matrix, i, j, matrix[i][j], False)
                if test == False:
                    return False
    return True
                
def test_cell(matrix, row, column, value, solvingCheck):
    #'solvingCheck' is true when solving and false when validating
    #it is used to skip current cell when validating
    #test row
    for i in range(9):
        if i == column and solvingCheck == False:
            continue
        if matrix[row][i] == value:
            return False
    #test column
    for i in range(9):
        if i == row and solvingCheck == False:
            continue
        if matrix[i][column] == value:
            return False    
    #test neighborhood
    x_position = row // 3
    y_position = column // 3
    for i in range(x_position*3, x_position*3+3):
        for j in range(y_position*3, y_position*3+3):
            if i == row and j == column:
                continue
            if matrix[i][j] == value:
                return False
    return True

def print_matrix(matrix):
    for i in range(len(matrix)):
        for j in range(len(matrix[i])):
            print(matrix[i][j], end=" ")
        print("",end="\n")
    return 
    
if __name__ == '__main__':
    main()